### PR TITLE
Starts work for keyless auth for robot accounts

### PR DESCRIPTION
**Pull request by stevsmit**

### CODE REVIEW

File: modules/keyless-authentication-robot-accounts.adoc [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+.. Click *Robot Accounts*.
+
+.. Click the menu kebab -> *Set robot federation*.
+
+.. Click the *+* symbol. 
+
+.. In the popup window, include the following information:
++
+* *Issuer URL*: `\https://keycloak-auth-realm.quayadmin.org/realms/quayrealm`, or the URL of your Keycloak realm.
+* *Subject*: `449e14f8-9eb5-4d59-a63e-b7a77c75f770`, or the *ID* of your keycloak user.
```
Comment (LiZhang19817) [inline modules/keyless-authentication-robot-accounts.adoc:141]: Here need to use the subject, can't use the **ID** of the user.

File: modules/keyless-authentication-robot-accounts.adoc [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+[source,text]
+----
+http://localhost:8080/realms/master/protocol/openid-connect/token
+----
+
+. On a web browser, navigate to the following URL:
++
+[source,text]
+----
+http://<keycloak_url>/realms/<realm_name>/protocol/openid-connect/auth?response_type=code&client_id=<client_id>
```
Comment (LiZhang19817) [inline modules/keyless-authentication-robot-accounts.adoc:57]: The example is not correct, need to add **"auth"**
https://sso-sso.apps.quaytest-3218.qe.azure.devcluster.openshift.com/**auth**/realms/quay/protocol/openid-connect/auth?response_type=code&client_id=quay

File: modules/keyless-authentication-robot-accounts.adoc [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+.Procedure
+
+. On the Keycloak main page:
+
+.. Select the appropriate realm that is authenticated for use with {productname}. Store the issuer URL, for example, `\https://keycloak-auth-realm.quayadmin.org/realms/quayrealm`.
+
+.. Click *Users* -> the name of the user to be linked with the robot account for authentication. You must use the same user account that you used when generating the OAuth2 access token.
+
+.. On the *Details* page, store the *ID* of the user, for example, `449e14f8-9eb5-4d59-a63e-b7a77c75f770`.
```
Comment (LiZhang19817) [inline modules/keyless-authentication-robot-accounts.adoc:126]: Actually need to input the **Subject** of the OIDC user, for keyCloak the ID on the KeyCloak UI is the returned subject of the user, but for Azure Entra ID the ID is not the Subject
**Example:**
"sub": "1b4f494a-496f-412f-a1b7-8db236d6ecbd",

File: modules/keyless-authentication-robot-accounts.adoc [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+----
+import requests
+import os
+
+TOKEN=os.environ.get('TOKEN')
+robot_user = "fed-test+robot1"
+
+def get_quay_robot_token(fed_token):
+    URL = "https://<quay-server.example.com>/oauth2/federation/robot/token"
+    response = requests.get(URL, auth=(robot_user,fed_token))
```
Comment (LiZhang19817) [inline modules/keyless-authentication-robot-accounts.adoc:173]: Here in this line code, when Quay is using the self-signed SSL Cert, need add one more parameter **verify=False**

`response = requests.get(URL,auth=(robot_user,federation_token),verify=False)`

File: modules/keyless-authentication-robot-accounts.adoc (hunk 2 above)
Comment (stevsmit) [reply modules/keyless-authentication-robot-accounts.adoc:57]: @LiZhang19817 just to make sure:

I got this URL from OpenID Endpoint Configuration under realm settings. See image for URL. 

Then I navigated to that using the ?response_type=code&client_id=<client_id> metadata. See image for UI endpoint. 

openid-auth

code-example

In this example, URL is https://localhost:3000/cb?session_state=a81c3488-8eb2-481e-a635-4abb0605303f&iss=http%3A%2F%2Flocalhost%3A8080%2Frealms%2Fmaster&code=0c7dbb1a-584c-4b33-a1f8-9bf90d147051.a81c3488-8eb2-481e-a635-4abb0605303f.cdffafbc-20fb-42b9-b254-866017057f43. The code is what i use with curl -x POST in the following command. 

I'm happy to change to https://sso-sso.apps.quaytest-3218.qe.azure.devcluster.openshift.com/auth/realms/quay/protocol/openid-connect/auth?response_type=code&client_id=quay which might be better? Thanks.

File: modules/keyless-authentication-robot-accounts.adoc (hunk 1 above)
Comment (stevsmit) [reply modules/keyless-authentication-robot-accounts.adoc:141]: Sorry, I meant "which is the *ID* of your keycloak user. not one or the other. sorry for the confusion.

File: modules/keyless-authentication-robot-accounts.adoc (hunk 4 above)
Comment (stevsmit) [reply modules/keyless-authentication-robot-accounts.adoc:173]: Updated

File: modules/keyless-authentication-robot-accounts.adoc [hunk 5]
```diff
@@ -0,0 +1,268 @@
+:_content-type: PROCEDURE
+[id="keyless-authentication-robot-accounts"]
+= Keyless authentication with robot accounts
+
+In previous versions of {productname}, robot account tokens were valid for the lifetime of the token unless deleted or regenerated. Tokens that do not expire have security implications for users who do not want to store long-term passwords or manage the deletion, or regeneration, or new authentication tokens. 
+
+With {productname} {producty}, {productname} administrators are provided the ability to exchange {productname} robot account tokens for an external OIDC token with either Red Hat SSO Key Cloak or Microsoft Entra ID. This allows robot accounts to leverage short-lived, or _ephemeral tokens_, that last one hour. Ephemeral tokens are refreshed regularly and can be used to authenticate individual transactions.
```
Comment (syed) [inline modules/keyless-authentication-robot-accounts.adoc:7]: it's the opposite. Exchange external OIDC tokens for a short lived robot token

File: modules/keyless-authentication-robot-accounts.adoc (hunk 3 above)
Comment (stevsmit) [reply modules/keyless-authentication-robot-accounts.adoc:126]: made a note.

File: modules/keyless-authentication-robot-accounts.adoc (hunk 5 above)
Comment (stevsmit) [reply modules/keyless-authentication-robot-accounts.adoc:7]: fixed ty

**Pull request merged**